### DISCUSSION TITLE
Alpine click handler not properly updated when livewire dom-diffs

### DISCUSSION THREAD

**Discussion by georgeboot** [livewire/livewire#1862] [1. Help]
### Description
Alpine click handler not properly updated when livewire dom-diffs

### Exact steps to reproduce
1. Go to https://laravelplayground.com/#/snippets/4a42ea39-627f-46d9-893e-41e69a33d7f8
2. Open your browsers console so you can see the logs
3. click the "increment" button a few timea. For each click, a "increment" log line will be written to the console
4. click the "toggle button 1" and now press the "decrement" button
5. The count will **not** go down by 1, it will instead go up by 1. However, a "decrementing" log will be written to the console.

Same for the variant with a stacks (instead of counter) and **with wire:key**.
1. Click the "stack1" button a few times. One foo element is added to the stack1 for each click.
2. Click the "toggle button 2". Now click the "stack 2" button a few times
3. For each click on "stack 2", **2 bar items** are added to the stack, instead of 1 bar item

**Context**
- Livewire version: 2.2.9
- Laravel version: 8.8.0
- Browser: Chrome, Firefox

**Comment by iksaku** [comment]
For your first issue with the `increment`/`decrement` buttons, you can easily fix it using the `wire:key` attribute, which will allow Livewire to know which instance of which its handling. This way, the DOM changes will properly be reflected.

For the second issue, well, that's a real issue. Not sure at the moment how to fix it, but we should investigate further on it.

**Comment by georgeboot** [comment]
@iksaku I mainly provided both examples to demonstrate they both with ánd without wire:key, it isn't handled properly.

If we make the elements a bit more complex, livewire should have no issues fpm-diffing it, and it works on many other places too.

I wonder if one should always set a wire:key to elements that also have Alpine actions attached to them? I can't think that's required as that would be clearly stated in the docs, would that be the case.

**Comment by iksaku** [comment]
Yeah, I see what you mean now. But still, this could be related to the way Alpine reacts to Livewire's DOM changes.

We should open investigation on the Alpine code and, if my assumptions are correct, then head over to the alpine repo with this issue and submit a PR.

**Comment by iksaku** [comment]
@ryangjchandler may I request your support for this issue?

**Comment by ryangjchandler** [comment]
@iksaku Won't be able to take a look tonight unfortunately.

**Comment by squishythejellyfish** [comment]
👋 Oh Hi! I'm Squishy, the friendly jellyfish that manages Livewire issues.

I see this issue has been closed.

Here in the Livewire repo, we have an "issues can be closed guilt-free and without explanation" policy.

If for ANY reason you think this issue hasn't been resolved, PLEASE feel empowered to re-open it.

Re-opening actually helps us track which issues are a priority.

Reply "REOPEN" to this comment and we'll happily re-open it for you!

(More info on this philosophy here: https://twitter.com/calebporzio/status/1321864801295978497)

**Comment by georgeboot** [comment]
@ryangjchandler @iksaku can one of you guys confirm this is indeed a LiveWire issue (too)? I have too little understanding of the code base to figure this out myself.

**Comment by iksaku** [comment]
I'm not Deep enough into the JS side of Livewire+Alpine, so I can't confirm that this is indeed a Livewire bug, as it could also be addressed somehow in Alpine.

I would like however to leave this open for further research.

@danharrin could we convert this into a Discussion so more people can find it and provide further discussion if possible?

**Reply by danharrin** [reply]
All done, make sure to open an issue if it's reproducible 👍

**Reply by georgeboot** [reply]
@danharrin Its already reproducible right? There is a playground link demonstrating the issue.

**Comment by joshhanley** [comment]
I believe this issue #763 is related. I've experienced the same issue myself over the last couple of days.

I've tried quickly debugging it, but I haven't given it a good go. Found a work around which I posted on the issues thread.

What was happening though, was any click handlers that are on an object inside a blade conditional or loop, work fine the first load, but if there are any changes as part of a livewire request, then when the new content is diffed into the page, Alpine seems to be doubling up on the click handling, calling each twice whenever one click happens.

**Reply by georgeboot** [reply]
That definitely seems related yes
